tests/test_cases/test_array_buses/test_array_buses.py

```diff
@@ -0,0 +1,79 @@
+import cocotb
+from cocotb.clock import Clock
+from cocotb.drivers import BusDriver
+from cocotb.monitors import BusMonitor
+from cocotb.triggers import RisingEdge
+
+
+class TestDriver(BusDriver):
+    _signals = ["data", "valid"]
+
+    def __init__(self, entity, name, clock, **kwargs):
+        BusDriver.__init__(self, entity, name, clock, **kwargs)
+        self.bus.valid <= 0
+
+    @cocotb.coroutine
+    def _driver_send(self, transaction, sync=True):
+        clkedge = RisingEdge(self.clock)
+        if sync:
+            yield clkedge
+
+        self.log.info("Sending {}".format(transaction))
+        self.bus.valid <= 1
+        self.bus.data <= transaction
+
+        yield clkedge
+        self.bus.valid <= 0
+
+
+class TestMonitor(BusMonitor):
+
+    _signals = ["data", "valid"]
+
+    def __init__(self, entity, name, clock, bank=0, **kwargs):
+        BusMonitor.__init__(self, entity, name, clock, **kwargs)
+        self.bank = bank
+        self.add_callback(self._get_result)
+        self.expected = []
+
+    @cocotb.coroutine
+    def _monitor_recv(self):
+        clkedge = RisingEdge(self.clock)
+        while True:
+            yield clkedge
+            if self.bus.valid.value:
+                self._recv(int(self.bus.data))
+
+    def _get_result(self, transaction):
+        self.log.info("Received {} on bank {}".format(transaction, self.bank))
+        exp = self.expected.pop(0)
+        if exp != int(transaction):
+            raise cocotb.result.TestFailure("Expected {}, got {}".format(exp, int(transaction)))
```

Comment:
If `_get_result` is never called (eg, the clock isn't actually working), then your test will still pass - you also need to check that `expected` is empty at the end of the main coroutine.

Also, shorter as:

```suggestion
    def _get_result(self, transaction):
        self.log.info("Received {} on bank {}".format(transaction, self.bank))
        exp = self.expected.pop(0)
        assert exp == int(transaction)
```